frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/UploadImageModel.java

```diff
@@ -161,30 +160,32 @@ public void setTestResponse(EntityModel<Response> testResponse) {
     public UploadImageModel(final Guid limitToStorageDomainId, final DiskImage resumeUploadDisk) {
         if (resumeUploadDisk == null) {
             setDiskModel(new NewDiskModel() {
+                private final boolean isChangeable = limitToStorageDomainId == null;
+
                 @Override
                 public void initialize() {
                     super.initialize();
 
-                    getStorageDomain().setIsChangeable(limitToStorageDomainId == null);
-                    getDataCenter().setIsChangeable(limitToStorageDomainId == null);
+                    getDataCenter().setIsChangeable(isChangeable);
+                    if (!isChangeable) {
+                        initStorageDomain();
+                    }
+                    getStorageDomain().setIsChangeable(isChangeable);
                     getStorageType().setIsChangeable(false);
                     getSize().setIsChangeable(false);
                 }
 
+                private void initStorageDomain() {
```

Comment:
`initStorageDomain()` is async and is executed (possibly) in parallel to `super.initialization()` flow. This looks as race condition to me. In order to be 100% safe this approach requires some kind of `onInitializationFinished()` callback that would guarantee that storage domain will be set **after** all initializations are done.

Reply:
but why do we care that it will be executed after all the initialization is completed if that initialization cannot change the storage domain?

Reply:
@sleviim 
OK. I had a test run and this version definitely is better then before (although not ideal from my perspective). Please resolve my comments - for some reasons I'm not able to do that.

Some thoughts:
1. running both queries (sd and dc) in parallel works in happy path (thanks to blocking `updateStorageDomains()`). If one of the queries would fail then we have partial state. However this worked this way before (no callback is [handling failure](https://github.com/oVirt/ovirt-engine/blob/f3cf526a05d5b414800ba3e1fd7bcb3438eeac91/frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/Model.java#L655))
2.  `getStorageDomain().getItems() == null` since it was never populated. However again this used to work this way before.
3. the DC seems selected based on [this condition](https://github.com/oVirt/ovirt-engine/blob/f3cf526a05d5b414800ba3e1fd7bcb3438eeac91/frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/vms/AbstractDiskModel.java#L462) - please check if you are OK with that

Reply:
@ahadas 
> but why do we care that it will be executed after all the initialization is completed if that initialization cannot change the storage domain?
>

The initialization can fail or may overwrite our changes. In the current code it seems we managed to block updating storage domain but this may change with any change in the super class.
Btw note that "freezing" the storage domain is only a way to "freeze" hosts. However both are actually not related. The hosts update depends from `performUpdateHosts()` extension point.